internal/main.go

```diff
@@ -0,0 +1,148 @@
+		}
+	}
+	wg.Wait()
+	logger.Info("pulling images finished")
+	return listImagesForDebugging(ctx, logger, client, timing.ImageListTimeout, "after")
+}
+
+func listImagesForDebugging(ctx context.Context, logger *slog.Logger, client criV1.ImageServiceClient, timeout time.Duration, stage string) error {
+	if !logger.Enabled(ctx, slog.LevelDebug) {
+		return nil
+	}
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+	logger.DebugContext(ctx, "starting to list images")
+	imagesResp, err := client.ListImages(ctx, &criV1.ListImagesRequest{})
+	if err != nil {
+		return err
+	}
+	logger.DebugContext(ctx, "finished listing images")
+	for _, i := range imagesResp.Images {
+		logger.DebugContext(ctx, "image present in runtime", "image", i, "stage", stage)
+	}
+	return nil
+}
+
+func loadPullSecret(logger *slog.Logger, kr *credentialprovider.BasicDockerKeyring, dockerConfigJSONPath string) error {
+	if dockerConfigJSONPath == "" {
+		logger.Info("no image pull secret path provided, will pull without credentials")
+		return nil
+	}
+	f, err := os.ReadFile(dockerConfigJSONPath)
+	if err != nil {
+		return err
+	}
+	dockerConfigJSON := credentialprovider.DockerConfigJSON{}
+	if err := json.Unmarshal(f, &dockerConfigJSON); err != nil {
+		return fmt.Errorf("unmarshalling docker config failed: %w", err)
+	}
+	kr.Add(dockerConfigJSON.Auths)
+	return nil
+}
+
+func getAuthsForImage(ctx context.Context, logger *slog.Logger, kr credentialprovider.DockerKeyring, imageName string) []*criV1.AuthConfig {
+	credsList, _ := kr.Lookup(imageName)
+	var auths []*criV1.AuthConfig
+	if len(credsList) == 0 {
+		logger.DebugContext(ctx, "no credentials present for image", "image", imageName)
+		// un-authenticated pull
+		auths = append(auths, nil)
+	}
+	for _, creds := range credsList {
+		auth := &criV1.AuthConfig{
+			Username:      creds.Username,
+			Password:      creds.Password,
+			Auth:          creds.Auth,
+			ServerAddress: creds.ServerAddress,
+			IdentityToken: creds.IdentityToken,
+			RegistryToken: creds.RegistryToken,
+		}
+		auths = append(auths, auth)
+	}
+	return auths
+}
+
+func pullImageWithRetries(ctx context.Context, logger *slog.Logger, wg *sync.WaitGroup, client criV1.ImageServiceClient, request *criV1.PullImageRequest, timing TimingConfig) {
+	defer wg.Done()
+	attemptTimeout := timing.InitialPullAttemptTimeout
+	delay := timing.InitialPullAttemptDelay
+	for {
+		logger.Info("attempting image pull", "timeout", attemptTimeout)
+		attemptCtx, cancel := context.WithTimeout(ctx, attemptTimeout)
+		start := time.Now()
+		response, err := client.PullImage(attemptCtx, request)
+		elapsed := time.Since(start)
+		cancel()
+		if err == nil {
+			logger.InfoContext(ctx, "image pulled successfully", "response", response, "elapsed", elapsed)
+			return
+		}
+		logger.ErrorContext(ctx, "image failed to pull", "error", err, "timeout", attemptTimeout, "elapsed", elapsed)
```

Comment:
Are there any errors that we should not retry? E.g. unauthenticated?

Reply:
I remember quay sometimes incorrectly claiming lack of privileges in the past, so I'd definitely want to retry auth errors.
In general I think this should be fine, given we have exponential backoff.